rchain-coop/src/routes/api.js

```diff
@@ -0,0 +1,82 @@
+var express = require('express');
+var router = express.Router();
+var request = require('request');
+var apiCache = require('apicache')
+
+
+const MAINNET_OBSERVER = "https://observer.services.mainnet.rchain.coop"
+// the total supply of the rchain mainnet is fixed
+const TOTAL_SUPPLY = 870663574.00
+const REV_TO_PHLO = 100000000
+const CIRCULATION_CACHE_TIMEOUT = '1 day'
+const BALANCE_CACHE_TIMEOUT = '5 minutes'
+
+
+// this registry result is put by the developer in the coop team
+// more details of the result is from https://gist.github.com/tgrospic/8355163bc1f84ed5e6e0af45b897ca8b
+const totalCirculationQuery = `
+new return, totalCh, lookup(\`rho:registry:lookup\`) in {
+  lookup!(\`rho:id:z3kotpcwux8ekqb85sdnmj9kj3mwyc1wpoqa1jwc1zcoqc9ptrn9ti\` , *totalCh) |
+    for (totalCirculation <- totalCh) {
+      totalCirculation!(*return)
+    }
+  }
+`
+
+function balanceQuery (address) {
+  var query = `
+new return, rl(\`rho:registry:lookup\`), RevVaultCh, vaultCh, balanceCh in {
+  rl!(\`rho:rchain:revVault\`, *RevVaultCh) |
+  for (@(_, RevVault) <- RevVaultCh) {
+    @RevVault!("findOrCreate", "$addr", *vaultCh) |
+    for (@(true, vault) <- vaultCh) {
+      @vault!("balance", *balanceCh) |
+      for (@balance <- balanceCh) {
+        return!(balance)
+      }
+    }
+  }
+}`
+  return query.replace("$addr", address)
+}
+
+router.get('/total-supply', function (req, res, next) {
+  res.json(TOTAL_SUPPLY);
+});
+
+router.get('/total-circulation', apiCache.middleware(CIRCULATION_CACHE_TIMEOUT), function (req, res, next) {
+  request.post(MAINNET_OBSERVER + '/api/explore-deploy', { body: totalCirculationQuery }, (err, resp, body) => {
+    if (resp.statusCode >= 400) {
+      res.send(resp.statusCode)
+    } else {
+      try {
+        const result = JSON.parse(body)
+        const circulation = result['expr'][0]['ExprInt']['data'] / REV_TO_PHLO
+        res.json(circulation)
+      }
+      catch{
+        res.send(400)
+      }
+    }
+  })
+});
+
+router.get('/balance/:address', apiCache.middleware(BALANCE_CACHE_TIMEOUT), function (req, res, next) {
```

Comment:
Do we want to be able to check balance only for coop addresses or for any address? I was thinking to create only what is necessary for now.

Reply:
I can not answer to that question.